### DISCUSSION TITLE
Error in user-defined CallBack loss and jacobian function. Got an error message with "Notify the CasADi developers."

### DISCUSSION THREAD

**Discussion by Tian-A-Tan** [casadi/casadi#3270] [Q&A] [answered]
Hi!

I'm using Casadi and a 3rd-party library for trajectory optimization. The loss and jacobian (loss w.r.t. x) are implicitly provided by the 3rd-party library so there is no symbolic loss or jacobian function. I implemented a custom LossFun with a JacFun following the official CallBack example and got the following error message.

I don't understand why it's attempting to compute reverse even though a jacobian is provided.

Here is the error message:
Traceback (most recent call last):
  File "/mnt/d/Local/AddBiom/vmu-suit/check.py", line 88, in <module>
    F = nlpsol('F', 'ipopt', nlp, dict(enable_reverse=False, enable_forward=False))
  File "/home/alan/.local/lib/python3.8/site-packages/casadi/casadi.py", line 28956, in nlpsol
    return _casadi.nlpsol(*args)
RuntimeError: .../casadi/core/function_internal.cpp:145: Error calling IpoptInterface::init for 'F':
Error in Function::factory for 'nlp' [MXFunction] at .../casadi/core/function.cpp:1758:
Failed to create nlp_grad:[x, p, lam:f, lam:g]->[f, g, grad:gamma:x, grad:gamma:p] with {"gamma": [f, g]}:
.../casadi/core/factory.hpp:584: Gradient generation failed:
Error in MX::gradient at .../casadi/core/generic_matrix.hpp:1468:
Error in MX::jtimes at .../casadi/core/generic_matrix.hpp:1456:
Error in MX::reverse at .../casadi/core/mx.cpp:1788:
Error in MXFunction::ad_reverse at .../casadi/core/mx_function.cpp:1101:
Error in MX::ad_reverse for node of type N6casadi4CallE at .../casadi/core/mx.cpp:2310:
Error in Call::ad_reverse for 'f' [CallbackInternal] at .../casadi/core/casadi_call.cpp:149:
.../casadi/core/function_internal.cpp:3028: Assertion "v.size() == nadj" failed:
Notify the CasADi developers.

Below is a simplified version of my code.

Best,
Alan

```Python
from casadi import *


class ThirdPartyLib:
    @staticmethod
    def loss_computation(q, dq, ddq):
        weights = [1, 1, 1]
        loss = sum(weights[0] * q**2 + weights[1] * dq**2 + weights[2] * ddq**2)
        return loss

    @staticmethod
    def jac_computation(q, dq, ddq, loss):
        weights = [1, 1, 1]
        jac = 2 * (weights[0] * q + weights[1] * dq + weights[2] * ddq)
        return jac


class LossFun(Callback):
    def __init__(self, name, data_len, opts={}):
        Callback.__init__(self)
        self.data_len = data_len
        self.construct(name, opts)

    # Number of inputs and outputs
    def get_n_in(self): return 1

    def get_n_out(self): return 1

    def get_sparsity_in(self, i):
        return Sparsity.dense(self.data_len)

    def get_sparsity_out(self, i):
        return Sparsity.dense(1, 1)

    # Initialize the object
    def init(self):
        print('initializing object')

    # Evaluate numerically
    def eval(self, arg):
        q, dq, ddq = vertsplit(arg[0], 333)
        loss = ThirdPartyLib.loss_computation(q, dq, ddq)
        return [DM(loss)]

    def has_jacobian(self): return True

    def get_jacobian(self, name, inames, onames, opts):
        class JacFun(Callback):
            def __init__(self, data_len, opts={}):
                Callback.__init__(self)
                self.data_len = data_len
                # opts["enable_reverse"] = False
                # opts["enable_forward"] = False
                self.construct(name, opts)

            def get_n_in(self): return 2

            def get_n_out(self): return 1

            def get_sparsity_in(self, i):
                if i == 0:
                    return Sparsity.dense(self.data_len)
                elif i == 1:  # nominal output
                    return Sparsity.dense(1, 1)

            def get_sparsity_out(self, i):
                return Sparsity.dense(self.data_len)

            def eval(self, arg):
                q, dq, ddq = vertsplit(arg[0], 333)
                loss = arg[1]
                jac = ThirdPartyLib.jac_computation(q, dq, ddq, loss)
                jac = DM(jac)
                return [jac]

        self.jac_callback = JacFun(self.data_len)
        return self.jac_callback


if __name__ == '__main__':
    data_len = 999
    x = MX.sym('x', data_len)
    nlp = {}
    nlp['x'] = x
    objective = LossFun('f', data_len)
    nlp['f'] = objective(x)
    F = nlpsol('F', 'ipopt', nlp, dict(enable_reverse=False, enable_forward=False))
```

**Comment by jgillis** [comment, marked as answer]
This definitely needs a better error message.
The issue is the shape of the returned Jacobian.
Should be `Sparsity.dense(1,self.data_len)`, not `Sparsity.dense(self.data_len,1)`.

A gradient (reverse mode pass) is requested from top-level. CasADi will use your Jacobian to emulate a reverse mode pass.
